Win32-network/src/System/Win32/Async/IOManager.hsc

```diff
@@ -0,0 +1,274 @@
+      let gqcsOverlappedIsNull = lpOverlapped == nullPtr
+
+      -- gqcsIODataPtr was allocated by 'withIODataPtr' or
+      -- 'wsaWaitForCompletion', and we are responsible to deallocate it but
+      -- only if 'gqcsCompletionKey' is 'True'.
+
+      if | gqcsOverlappedIsNull && errorCode == eRROR_ABANDONED_WAIT_0 ->
+           -- this is triggered by closing IO completion 'HANDLE' while there
+           -- are outstanding requests (this is documented
+           -- <https://docs.microsoft.com/en-us/windows/win32/api/ioapiset/nf-ioapiset-getqueuedcompletionstatus#remarks>)
+              return ()
+         | gqcsOverlappedIsNull && errorCode == eRROR_INVALID_HANDLE ->
+           -- If we don't terminate the dequeueing thread, some of the tests
+           -- will take 10x-100x more time to complete:
+           --
+           -- - 'Async.Handle.async reads and writes'
+           -- - 'Async.Handle.PingPongPipelined test'
+           -- - 'Async.Socket.PingPong test'
+           -- - 'Async.Socket.PingPongPipelined test'
+           --
+           -- or not terminate at all within reasonable timeout:
+           --
+           -- - 'Async.Handle.PingPong test'
+           --
+           -- Note that in the test setup, we start io manager thread for each
+           -- test case.  If we don't terminate it when the iocp port is closed
+           -- we might end up with a busy loop which runs expensive system
+           -- calls that fail.  This could explain the slowdown.
+           --
+           -- In every observed case both values:
+           --
+           -- - 'gqcsResult'
+           -- - 'gqcsCompletionKey'
+           --
+           -- where false.
+           --
+               return ()
+         | gqcsOverlappedIsNull ->
+           -- 'GetLastCompletionStatus' has not dequeued any completion packet
+           -- from the completion port.  Must be the first clause, since if
+           -- this is not true we cannot trust other arguments.
+              throwIO NullOverlappedPointer
+         | completionKey /= magicCompletionKey ->
+            dequeueCompletionPackets iocp
+         | gqcsResult -> do
+           -- 'GetQueuedCompletionStatus' system call returned without errors.
+             !(numBytes :: Int) <-
+                 fromIntegral <$> peek numBytesPtr
+             let -- we can cast @Ptr OVERLAPPED@ to @Ptr AsyncIOCPData@ since
+                 -- 'OVERLAPPED' is a frist member of '_IODATA' struct.
+                 ioDataPtr :: Ptr AsyncIOCPData
+                 ioDataPtr = castPtr lpOverlapped
+             mvarPtr <- peek (iodDataPtr AsyncSing ioDataPtr)
+             mvar <- deRefStablePtr mvarPtr
+             freeStablePtr mvarPtr
+             free ioDataPtr
+             success <- tryPutMVar mvar (Right numBytes)
+             when (not success)
+               $ fail "System.Win32.Async.dequeueCompletionPackets: MVar is not empty."
+             dequeueCompletionPackets iocp
```

Comment:
Wait! We cannot call ourselves recursively here. You're intending this to be a loop (tail recursion), but it is not. You are still inside the `alloca` at the top, so you are leaking both heap memory and the stack.

Do the `alloca` and peek the results within the scope of the `alloca`s. But then do the various case analysis and acting on those cases outside the scope of the allocs. It might be clearer if the FFI bits are in a separate local function (so that it's really clear all the `peek`s are done and the allocations have been cleaned up), and the recursive worker function that does the case analysis is in another.

Reply:
Very good point.